feat(swarm): Make executor for connection tasks explicit (#3097)
paritytech/rust-libp2p: swarm/CHANGELOG.md

- Export `NetworkBehaviour` derive as `libp2p_swarm::NetworkBehaviour`.
  This follows the convention of other popular libraries. `serde` for example exports the `Serialize` trait and macro as
  `serde::Serialize`. See [PR 3055].

- Feature-gate `NetworkBehaviour` macro behind `macros` feature flag. See [PR 3055].

> - Made Swarm executor aware. See [PR 3097] for details.

Reviewer: ```suggestion
- Make executor in Swarm constructor explicit. See [PR 3097].
```

> - Make executor in Swarm constructor explicit. See [PR 3097].

Reviewer 2: Thank you for the great changelog entry.

Author: Thanks :)

  
  Supported executors:
  - Tokio

    Previously
    ```rust
    let swarm = SwarmBuilder::new(transport, behaviour, peer_id)
        .executor(Box::new(|fut| {
                tokio::spawn(fut);
        }))
        .build();
    ```
    Now
    ```rust
    let swarm = Swarm::with_tokio_executor(transport, behaviour, peer_id);

>     ```

Reviewer: Would this look better if we used a markdown table with two columns?

Author: Maybe, but it doesn't really work with code highlighting and `<br>`. So we probably have to leave it like this.

Reviewer: I see. That is a shame. Thanks for trying!

  - Async Std
    
    Previously
    ```rust
    let swarm = SwarmBuilder::new(transport, behaviour, peer_id)
        .executor(Box::new(|fut| {
                async_std::task::spawn(fut);
        }))
        .build();
    ```
    Now
    ```rust
    let swarm = Swarm::with_async_std_executor(transport, behaviour, peer_id);
    ```
  - ThreadPool (see [Issue 3107])
    
    In most cases ThreadPool can be replaced by executors or spawning on the local task.

    Previously
    ```rust
    let swarm = Swarm::new(transport, behaviour, peer_id);
    ```

    Now

>     ```

Reviewer: ```suggestion
    ```rust
```

    let swarm = Swarm::with_threadpool_executor(transport, behaviour, peer_id);
    ```
  - Without
    

>     Spawns the tasks on the current task, this results in bad performance so try to use an executor if possible. Previously this was just a fallback for ThreadPool.

Reviewer: ```suggestion
    Spawns the tasks on the current task, this may result in bad performance so try to use an executor where possible. Previously this was just a fallback when no executor was specified and constructing a `ThreadPool` failed.
```
